### DISCUSSION TITLE
Enable 32 bit Object Layer filtering

### DISCUSSION THREAD

**Discussion by captainkidd5** [amerkoleci/JoltPhysicsSharp#93] [Q&A] [answered]
looking [here](https://github.com/amerkoleci/JoltPhysicsSharp/blob/428027f5208a3bf5411813c1b21bb6f57481de30/src/JoltPhysicsSharp/ObjectLayer.cs#L8), is this currently hardcoded? If I want to change this to be int32 should I just be forking the repo and changing the source? I'm not really familiar with C++ but I'm assuming that Jolt lets you change that in a config file somewhere right?

I also notice that [this](https://github.com/amerkoleci/JoltPhysicsSharp/blob/428027f5208a3bf5411813c1b21bb6f57481de30/src/samples/HelloWorld/Program.cs#L209) line is commented out, which appears to be the bitmasking alternative implementation to using the mapping table for collision layers. The signature asks for a uint32, does this override the objectlayers settings of 16 bit? 

Edit - Looking at the api documentation - 

> [ObjectLayerPairFilterMask](https://jrouwe.github.io/JoltPhysics/class_object_layer_pair_filter_mask.html), [ObjectVsBroadPhaseLayerFilterMask](https://jrouwe.github.io/JoltPhysics/class_object_vs_broad_phase_layer_filter_mask.html) and [BroadPhaseLayerInterfaceMask](https://jrouwe.github.io/JoltPhysics/class_broad_phase_layer_interface_mask.html): These split an ObjectLayer in an equal amount of bits for group and mask. Two objects collide if (object1.group & object2.mask) != 0 && (object2.group & object1.mask) != 0. This behavior is similar to e.g. Bullet. In order to map groups to broad phase layers, you call [BroadPhaseLayerInterfaceMask::ConfigureLayer](https://jrouwe.github.io/JoltPhysics/class_broad_phase_layer_interface_mask.html#a01a9ec0c57462e3cda401b73edbf3573) for each broad phase layer. You determine which groups can be put in that layer and which group must be excluded from that layer. E.g. a broad phase layer could include everything that has the STATIC group but should exclude everything that has the SENSOR group, so that if an object has both STATIC and SENSOR bits set, this broad phase layer will not be used. The broad phase layers are checked one by one and the first one that meets the condition is the one that the body will be put in. If you use this implementation, consider setting the cmake option OBJECT_LAYER_BITS to 32 to get a 32-bit ObjectLayer instead of a 16-bit one.

It looks we have the proper bindings, I'm just confused on how to wrangle the OBJECT_LAYER_BITS setting, and also am confused by 

[the purpose of BroadPhaseLayerInterface and nint](https://github.com/amerkoleci/JoltPhysicsSharp/blob/428027f5208a3bf5411813c1b21bb6f57481de30/src/JoltPhysicsSharp/BroadPhaseLayerInterface.cs)

which is looks like I need to implement myself in order to pass to 
`ObjectVsBroadPhaseLayerFilterMask`

**Comment by amerkoleci** [comment]
Hi,
Jolt has option for that:
https://github.com/amerkoleci/joltc/blob/main/CMakeLists.txt#L76

But I wouldn't like to expand from ushort to uint, why do you need this?

**Reply by amerkoleci** [reply, marked as answer]
I created and example with alternative filtering, take look here:

https://github.com/amerkoleci/JoltPhysicsSharp/blob/main/src/samples/HelloWorld/Samples/AlternativeCollissionFilteringSample.cs

**Reply by captainkidd5** [reply]
The sample looks great, thanks! As for ushort to uint, I guess I'm unsure if I should just be casting the uint to ushort....


For example, if I wanted to have the player be present in both of these groups:
 
`CharacterVirtual.ExtendedUpdate(dt, _extendedUpdateSettings,new ObjectLayer((ushort)(GROUP_PLAYER | GROUP_TEST)), Game1.PhysicsSim.PhysicsSystem);
`

**Reply by captainkidd5** [reply]
Ah, nevermind, I also overlooked the box creation example in there. This all looks fantastic, appreciate it!
